### PR TITLE
feat: add select-rich

### PR DESCRIPTION
**Pull request by daKmoR**
has multiple requirements that need to be met before this can be considered
- [x] Requirement 1: [fix(field): no DomHelpers for FormControl and apply all needed mixins](https://github.com/ing-bank/lion/pull/158)
- [x] Requirement 2: [fix(field): remove CssClassMixin from InteractionStateMixin](https://github.com/ing-bank/lion/pull/159)
- [x] Requirement 3: [fix(validate): remove deprecated CssClassMixin](https://github.com/ing-bank/lion/pull/160)
- [x] Requirement 4: fix(field): the FormatMixin should work without an input element [open TODO]

Until Requirements are met this will remain a draft
- a lot of stuff already works
- lots and lots of tests

=> put here to extract/discuss specifics things that should be extracted before merging this

### CODE REVIEW

File: packages/button/src/LionButton.js [hunk 1]
```diff
@@ -12,9 +12,13 @@ export class LionButton extends DelegateMixin(SlotMixin(LionLitElement)) {
         type: String,
         reflect: true,
       },
-      tabindex: {
-        type: Number,
+      // we us a property here as if we use the native tabIndex we can not set a default value
+      // in the constructor as it synchronously sets the attribute which is not allowed in the
```
Comment (daKmoR) [inline packages/button/src/LionButton.js]: @gerjanvangeest can you extract this change to LionButton into a separate MR - pretty please 🤗 
probably will need some adoptions to tests as well (or if not tested then it should be)

File: packages/field/src/FormControlMixin.js [hunk 2]
```diff
@@ -76,7 +76,31 @@ export const FormControlMixin = dedupeMixin(
       }
 
       get inputElement() {
-        return (this.$$slot && this.$$slot('input')) || this.querySelector('[slot=input]'); // eslint-disable-line
+        if (this.__inputElementCache) {
+          return this.__inputElementCache;
+        }
+        this.__inputElementCache = this.querySelector('[slot=input]');
+      }
```
Comment (daKmoR) [inline packages/field/src/FormControlMixin.js]: should be
```js
        if (!this.__inputElementCache) {
          this.__inputElementCache = this.querySelector('[slot=input]');
        }
        return this.__inputElementCache;
```

should also be extracted to a separate MR

File: packages/field/src/FormatMixin.js [hunk 3]
```diff
@@ -336,14 +336,19 @@ export const FormatMixin = dedupeMixin(
           // is guaranteed to be calculated
           setTimeout(this._reflectBackFormattedValueToUser);
         };
```
Comment (daKmoR) [inline packages/field/src/FormatMixin.js]: this is only adding a "guard" to the `this.inputElement`... so the FormatMixin also works when applied to classes that do not have an underlying native inputElement...

=> should be extracted into a separate MR

File: packages/field/src/InteractionStateMixin.js [hunk 4]
```diff
@@ -16,7 +16,7 @@ import { FocusMixin } from './FocusMixin.js';
 export const InteractionStateMixin = dedupeMixin(
   superclass =>
     // eslint-disable-next-line no-unused-vars, no-shadow
-    class InteractionStateMixin extends CssClassMixin(FocusMixin(ObserverMixin(superclass))) {
+    class InteractionStateMixin extends ObserverMixin(superclass) {
       static get properties() {
         return {
```
Comment (daKmoR) [inline packages/field/src/InteractionStateMixin.js]: removes the CssClassMixin as it "breaks" the prototype chain...

should use `reflect: true` for most properties and should for backward compatibility still add the css classes

File: packages/validate/src/ValidateMixin.js [hunk 5]
```diff
@@ -15,7 +15,7 @@ const pascalCase = str => str.charAt(0).toUpperCase() + str.slice(1);
 export const ValidateMixin = dedupeMixin(
   superclass =>
     // eslint-disable-next-line no-unused-vars, no-shadow, max-len
-    class ValidateMixin extends CssClassMixin(ObserverMixin(LocalizeMixin(SlotMixin(superclass)))) {
+    class ValidateMixin extends ObserverMixin(LocalizeMixin(SlotMixin(superclass))) {
       /* * * * * * * * * *
```
Comment (daKmoR) [inline packages/validate/src/ValidateMixin.js]: removes the CssClassMixin as it "breaks" the prototype chain...

should use reflect: true for most properties and should for backward compatibility still add the css classes

=> separate MR

File: packages/field/src/InteractionStateMixin.js [hunk 6]
```diff
@@ -70,12 +70,16 @@ export const InteractionStateMixin = dedupeMixin(
         return !!value || isNumberValue || isBooleanValue;
       }
 
+      get _leaveNode() {
```
Comment (daKmoR) [inline packages/field/src/InteractionStateMixin.js]: it's probably better to not do this

implementers should capture and refire the leave event on the host instead (also results in a better API for them as well)

File: packages/field/src/FormControlMixin.js [hunk 7]
```diff
@@ -76,7 +76,31 @@ export const FormControlMixin = dedupeMixin(
       }
 
       get inputElement() {
-        return (this.$$slot && this.$$slot('input')) || this.querySelector('[slot=input]'); // eslint-disable-line
+        if (!this.__inputElementCache) {
```
Comment (daKmoR) [inline packages/field/src/FormControlMixin.js]: improved version of it can be found at https://github.com/ing-bank/lion/pull/158

we should sync this code with that one...

File: packages/field/src/InteractionStateMixin.js (hunk 4 above)
Comment (daKmoR) [reply packages/field/src/InteractionStateMixin.js]: MR is here: https://github.com/ing-bank/lion/pull/159

File: packages/option/src/LionOption.js [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        }
+
+        :host([disabled]) {
+          color: #adadad;
+        }
+      `,
+    ];
+  }
+
+  static get disableMixinManagesTabIndex() {
```
Comment (daKmoR) [inline packages/option/src/LionOption.js]: remove

File: packages/option/test/lion-option.test.js [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    it('can be active', async () => {
+      const el = await fixture(html`
+        <lion-option .choiceValue=${10}></lion-option>
+      `);
+      expect(el.active).to.equal(false);
+      expect(el.hasAttribute('active')).to.be.false;
+
+      el.active = true;
+      expect(el.active).to.be.true;
+      expect(el.hasAttribute('active')).to.be.false;
```
Comment (daKmoR) [inline packages/option/test/lion-option.test.js]: separate test for `// check that dom update is async`

File: packages/option/src/LionOption.js [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    return false;
+  }
+
+  constructor() {
+    super();
+    this.operatingSystem = 'auto';
```
Comment (daKmoR) [inline packages/option/src/LionOption.js]: remove

File: packages/select-rich/src/LionOptions.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+      },
+    };
+  }
+
+  constructor() {
+    super();
+    this.role = 'listbox';
+    // TODO: make a real property?
+    // eslint-disable-next-line wc/no-constructor-attributes
+    this.tabIndex = 0;
```
Comment (daKmoR) [inline packages/select-rich/src/LionOptions.js]: remove?

File: packages/select-rich/src/LionSelectInvoker.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+      const labelNodes = Array.from(this.selectedElement.querySelectorAll('*'));
+      if (labelNodes.length > 0) {
+        return labelNodes.map(node => node.cloneNode(true));
+      }
+      return this.selectedElement.textContent;
+    }
+    return ``;
+  }
+
+  render() {
```
Comment (daKmoR) [inline packages/select-rich/src/LionSelectInvoker.js]: do not override full render but only the `_renderAfter`

File: packages/select-rich/test/lion-select-rich-interaction.test.js [hunk 13]
```diff
@@ -0,0 +1,638 @@
...
+          <lion-options slot="input">
+            <lion-option .choiceValue=${10}>Item 1</lion-option>
+            <lion-option .choiceValue=${20}>Item 2</lion-option>
+          </lion-options>
+        </lion-select-rich>
+      `);
+      expect(el.checkedValue).to.equal(10);
+    });
+
+    it('can not be navigated with keyboard if disabled', async () => {
```
Comment (jorenbroekema) [inline packages/select-rich/test/lion-select-rich-interaction.test.js]:
```suggestion
    it('cannot be navigated with keyboard if disabled', async () => {
```

File: packages/select-rich/test/lion-select-rich-interaction.test.js [hunk 14]
```diff
@@ -0,0 +1,638 @@
...
+      expect(el.checkedIndex).to.equal(0);
+
+      el._listboxNode.dispatchEvent(new KeyboardEvent('keyup', { key: 'ArrowUp' }));
+      expect(el.activeIndex).to.equal(0);
+      expect(el.checkedIndex).to.equal(0);
+    });
+  });
+
+  describe('Disabled', () => {
+    it('can not be focused if disabled', async () => {
```
Comment (jorenbroekema) [inline packages/select-rich/test/lion-select-rich-interaction.test.js]:
```suggestion
    it('cannot be focused if disabled', async () => {
```

File: packages/select-rich/test/lion-select-rich-interaction.test.js [hunk 15]
```diff
@@ -0,0 +1,638 @@
...
+      const el = await fixture(html`
+        <lion-select-rich disabled></lion-select-rich>
+      `);
+      expect(el._invokerNode.tabIndex).to.equal(-1);
+    });
+
+    it('still has an checked value', async () => {
```
Comment (jorenbroekema) [inline packages/select-rich/test/lion-select-rich-interaction.test.js]:
```suggestion
    it('still has a checked value', async () => {
```

File: packages/select-rich/test/lion-select-rich-interaction.test.js [hunk 16]
```diff
@@ -0,0 +1,638 @@
...
+          <lion-options slot="input">
+            <lion-option .choiceValue=${10} disabled>Item 1</lion-option>
+            <lion-option .choiceValue=${20}>Item 2</lion-option>
+          </lion-options>
+        </lion-select-rich>
+      `);
+      const options = [...el.querySelectorAll('lion-option')];
+      el.disabled = true;
+      await el.updateComplete;
+      expect(options[0].disabled).to.be.true;
```
Comment (daKmoR) [inline packages/select-rich/test/lion-select-rich-interaction.test.js]: add test to renable if disabled on start

File: packages/select-rich/src/LionSelectRich.js [hunk 17]
```diff
@@ -0,0 +1,611 @@
...
+  set activeIndex(index) {
+    if (this.formElements[index]) {
+      this.formElements[index].active = true;
+    }
+  }
+
+  constructor() {
+    super();
+    this.interactionMode = 'auto';
+    this.formElements = [];
```
Comment (daKmoR) [inline packages/select-rich/src/LionSelectRich.js]: already done in 'RegistrarMixin'

File: packages/option/test/lion-option.test.js (hunk 9 above)
Comment (daKmoR) [reply packages/option/test/lion-option.test.js]: @jorenbroekema renamed it to `adds asynchronous the attribute "active" when becomes active` as splitting the tests seemed to not help 🙈

File: packages/select-rich/test/lion-select-rich-interaction.test.js (hunk 16 above)
Comment (daKmoR) [reply packages/select-rich/test/lion-select-rich-interaction.test.js]: test added => and fixed

File: packages/option/test/lion-option.test.js (hunk 9 above)
Comment (jorenbroekema) [reply packages/option/test/lion-option.test.js]: Maybe rename to "asynchronously adds the 'active' attribute when the element is active" for grammar :P

Comment (daKmoR) [reply packages/option/test/lion-option.test.js]: changed

**Comment by daKmoR** [comment]
This is now ready 💪 🎉
